upgradability/forward-proxy/forward_proxy/lib.rs

```diff
@@ -0,0 +1,130 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+mod forward_proxy {
+    use ink_env as env;
+    use ink_env::call::Call;
+    use scale::{Decode, Encode};
+
+    #[derive(Eq, PartialEq, Debug, Decode, Encode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        PermissionDenied,
+        ContractCallError,
+    }
+
+    #[ink(storage)]
+    pub struct ForwardProxy {
+        logic_contract: AccountId,
+        admin: AccountId,
+    }
+
+    impl ForwardProxy {
+        #[ink(constructor)]
+        pub fn new(logic_contract: AccountId) -> Self {
+            Self {
+                logic_contract,
+                admin: Self::env().caller(),
+            }
+        }
+
+        /// Allows admin to transfer his privilages to another account
+        #[ink(message)]
+        pub fn change_admin(&mut self, new_admin: AccountId) -> Result<(), Error> {
+            if self.env().caller() != self.admin {
+                Err(Error::PermissionDenied)
+            } else {
+                self.admin = new_admin;
+                Ok(())
+            }
+        }
+
+        /// Allows admin to change address of the logic contract
+        #[ink(message)]
+        pub fn change_logic_contract(
+            &mut self,
+            new_logic_contract: AccountId,
+        ) -> Result<(), Error> {
+            if self.env().caller() != self.admin {
+                Err(Error::PermissionDenied)
+            } else {
+                self.logic_contract = new_logic_contract;
+                Ok(())
+            }
+        }
+
+        /// A catch-all method which forwards calls which selectors
+        /// do not match other methods of that proxy
+        #[ink(message, payable, selector = _)]
+        pub fn _catch_all_forward(&self) -> Result<(), Error> {
+            match env::call::build_call::<env::DefaultEnvironment>()
+                .call_type(
+                    Call::new()
+                        .callee(self.logic_contract)
+                        .transferred_value(self.env().transferred_value()),
+                )
+                .call_flags(
+                    env::CallFlags::default()
+                        .set_forward_input(true)
+                        .set_tail_call(true),
+                )
+                .fire()
+            {
+                Err(_) => Err(Error::ContractCallError),
```

Comment:
maybe don't ignore the why, it can have some useful context, if it's not encode/decode you can always .to_string it and return with your own contract's error variant (e.g. ContractCallError(String) )